train_gpt2.c

```diff
@@ -18,6 +18,18 @@ There will be other versions of this code that specialize it and make it fast.
 #include <omp.h>
 #endif
 
+#ifndef NDEBUG // Make zero cost when not debugging
```

Comment:
If the assumptions is that it's "fine" (as in process terminating/crashing) to deref a null pointer, then what's the point of handling in debug mode?

If anything, it makes more sense to swap it around. Because if you're debugging, you are (or should be) already running in a debugger and/or have Undefined Sanitizer enabled which will break on the null deref, rendering the error printout as useless. Whereas on release builds the printout would've been more useful.

Reply:
Not wrong. As I thought about this more maybe the callocs should not exist at all? Ideally there is one malloc() for an arena.